enhancements/baremetal-config.md

```diff
@@ -0,0 +1,127 @@
+
+### Non-Goals
+
+All config values except the RHCOSImageUrl are not expected to change after a
+metal3 cluster is up. Handling changes to this value after install time has
+not been explored and that is a non-goal at this time.
+
+### Proposal
+
+To support configuration of a Baremetal IPI (metal3) deployment, a set of new
+config items are required to be made availabe via a Config Resource. The
+BareMetalPlatformStatus is an ideal fit because it contains configuration
+that is specific to a Baremetal deployment.
+
+This proposal aims to add the following to BareMetalPlatformStatus:
+
+1. ProvisioningInterface : This is the interface name on the underlying
+baremetal server which would be physically connected to the provisioning
+network.
+
+2. ProvisioningIP : This is the IP address used to to bring up a NIC on
+the baremetal server for provisioning. This value should not be in the
+DHCP range and should not be used in the provisioning network for any
+other purpose (should be a free IP address.) It is expected to be provided
+in the format : 10.1.0.3.
+
+3. ProvisioningNetworkCIDR : This is the provisioning network with its
+CIDR. The ProvisioningIP and the ProvisioningDHCPRange are expected to
+be within this network. The value for this config is expected to be
+provided in the format : 10.1.0.0/24.
+
+4. ProvisioningDHCPRange : This is a range of IP addresses from which
+baremetal hosts can be assigned their IP address. This range should be
+provided as two comma seperated IP addresses where the addresses represent
+a contiguous range of IPs that are available to be allocated. It is expected
+to be provided in the format : 10.1.0.10, 10.1.0.100.
+
+5. CacheURL : This is a location where image required to bring up baremetal
+hosts have been previously downloaded for faster image downloads within the
+cluster. This is an optional parameter and is expected to be provided
+in the format : http://111.111.111.1/images
+
+6. RHCOSImageURL : This config is expected to be the location of the
+official RHCOS release image. It is a required parameter and will be used
+to deploy new nodes in the cluster.
+
+### User Stories [optional]
+
+This enhancement will allow users to have their Barametal IPI deployments
+customizable to their specific hardware and network requirements.
+
+## Design Details
+
+### Test Plan
+
+The test plan should involve making sure the openshift/installer generates
+all configuration items within the BaremetalPlatformStatus when the platform
+type is Baremetal.
+
+MAO reads this configuration and uses these to derive additional configuration
+required to bring up a metal3 cluster. E2e testing should make sure that MAO
+is able to bring up a metal3 cluster using config from this enhanced
+BaremetalPlatformStatus CR.
+
+### Upgrade / Downgrade Strategy
+
+This enhancement is applicable only to the Baremetal platform type. Since,
+this platform type is not officially supported in 4.2, we feel that an
```

Comment:
Please link to where in the docs we say this isn't supported and to where in the code you're preventing upgrades from happening.

Reply:
As far as I know, we are not preventing upgrades from happening in the code. The current upgrade strategy comes less from the code and more from business logic (for lack of a better word in my vocabulary). @russellb should be able to clarify this further.

Reply:
@deads2k this is a net-new integration for a currently unsupported/experimental platform so we don't expect there to be any users to upgrade, we'll only support new deployments in the first instance.

Perhaps you can expand a bit on the sort of strategy/docs you would expect in this case, clearly when this becomes fully supported we'll need to ensure we have a strategy to deal with any changes in this schema, but right now I don't think it's a concern, we're still trying to bootstrap this new baremetal IPI platform integration (and this API change is one of the final pieces to enable that)

Reply:
> @deads2k this is a net-new integration for a currently unsupported/experimental platform so we don't expect there to be any users to upgrade, we'll only support new deployments in the first instance.
> 
> Perhaps you can expand a bit on the sort of strategy/docs you would expect in this case, clearly when this becomes fully supported we'll need to ensure we have a strategy to deal with any changes in this schema, but right now I don't think it's a concern, we're still trying to bootstrap this new baremetal IPI platform integration (and this API change is one of the final pieces to enable that)

I just want to be sure that customers cannot possibly attempt to upgrade from 4.2 to 4.3.  If they can, the chances that they will go from a working cluster to a completely broken one are fairly high.  Simply saying, "this isn't supported" doesn't really satisfy, but Derek told me that you can't actually get thing to install in 4.2, so I'm no longer worried about it.

Reply:
Yeah, currently if anyone tries to deploy using the baremetal platform, the baremetal operator pod will simply crashloop on start because of the missing configmap, we've worked around it for developer testing by manually adding it, but I think for end-users it's extremely unlikely they would do such an upgrade, and also this platform is marked experimental in the installer.

FWIW that's also why I think this is a bug and not really an enhancement - the current MAO integration is broken without passing this data from the installer.

Reply:
> please add a discussion about how the information is stored today and how you'll migrate that data. If you don't support upgrade, please indicate that, link to the docs, and ideally where in the code you actually stop that upgrade.

This information is not currently stored anywhere. For development testing purposes, we have scripts that create a configmap that contains this config. MAO code in 4.2 refers to this ConfigMap which either has to be provided by these scripts (or manually?) when the platform type is "Baremetal". 
So, during the 4.3 cycle, we had 1st attempted to have the installer add the ConfigMap but that effort was abandoned in favor of adding these config items to the BaremetalPlatformStatus CR. Hopefully, the enhancement request and this discussion provides the necessary context for this effort.